Perfetto connector

jrmadsen opened the pull request:
- alternative implementation to https://github.com/kokkos/kokkos-tools/pull/124
  - Has some configuration options
  - Supports both in-process and system backends

jrmadsen committed:
Perfetto connector

dalg24 commented on profiling/perfetto-connector/libperfetto-connector.cpp:
```diff
@@ -0,0 +1,182 @@
+}
+
+void kokkosp_begin_parallel_for(const char* name, const uint32_t, uint64_t*) {
+  TRACE_EVENT_BEGIN("kokkos.kernels", perfetto::StaticString(name));
```
Not distinguishing `parallel_for` from `parallel_reduce` and `parallel_scan`?
I don't necessary have an issue with it.  Just wondering if that is by design.

jrmadsen replied:
Nope not by design, just inherited this from David's. It's a fair point though and easy to fix

dalg24 commented on profiling/perfetto-connector/libperfetto-connector.cpp:
```diff
@@ -0,0 +1,182 @@
+}
+
+void kokkosp_finalize_library() {
+  printf("-----------------------------------------------------------\n");
+  printf("KokkosP: Finalizing Perfetto Connector... ");
+  fflush(stdout);
```
Aww man I was so pleased with the environment parsing and now we are back to the good ole printf?

jrmadsen replied:
🤣 I just tweaked this from David's. I used to never use it but, over time, I've come to appreciate it more. I'd replace it with std::format in a heartbeat.

dalg24 commented on profiling/perfetto-connector/libperfetto-connector.cpp:
```diff
@@ -0,0 +1,182 @@
+
+template <typename Tp>
+auto get_env(std::string_view _env_name, Tp _default) {
+  const char* _env_val = std::getenv(_env_name.data());
```
I think that is potentially a bug.  `std::getenv` expects null-terminated character string.
I would pass a `std::string` and use `c_str()`.  This is not performance critical.

jrmadsen replied:
Good point

dalg24 reviewed (commented):
Looks mostly good besides the `string_view` thing

jrmadsen committed:
Address review, additional callbacks, debug annotations

- printf -> std::cout
- get_env uses std::string
- Renamed categories
- Debug annotations
- Support deep copy, alloc/dealloc data, profile evet, dual view

dalg24 commented on profiling/perfetto-connector/Makefile:
```diff
@@ -0,0 +1,18 @@
+
+default: all
+
+all: library strip
+
+library: libperfetto-connector.cpp
```
It is rebuilding each time
```suggestion
all: libperfetto-connector.so strip

libperfetto-connector.so: libperfetto-connector.cpp
```

dalg24 commented on profiling/perfetto-connector/Makefile:
```diff
@@ -0,0 +1,18 @@
+library: libperfetto-connector.cpp
+	$(CXX) $(CXXFLAGS) -I. libperfetto-connector.cpp perfetto.cc -shared -o libperfetto-connector.so
+
+strip: libperfetto-connector.so
+	strip -w -K kokkosp_* libperfetto-connector.so
```
Doesn't work on OSX
```
strip -w -K kokkosp_* libperfetto-connector.so
error: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/strip: unrecognized option: -w
Usage: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/strip [-AnuSXx] [-] [-d filename] [-s filename] [-R filename] [-o output] file [...]
make: *** [strip] Error 1
```

jrmadsen replied:
Fixed

dalg24 commented on profiling/perfetto-connector/libperfetto-connector.cpp:
```diff
@@ -0,0 +1,285 @@
+
+void kokkosp_finalize_library() {
+  std::cout << "-----------------------------------------------------------\n"
+            << "KokkosP: Finalizing Perfetto Connector... " << std::flush;
```
```suggestion
            << "KokkosP: Finalizing Perfetto Connector... " << std::endl;
```

```
17: -----------------------------------------------------------
17: KokkosP: Finalizing Perfetto Connector... [214.151] perfetto.cc:47237       Scraping SMB for producer 1
17: [214.156] perfetto.cc:38681       Stopping data source 1
17: [214.156] perfetto.cc:38721       Ending async stop of data source 1
17: [214.156] perfetto.cc:47237       Scraping SMB for producer 1
17: Reading trace data... Writing 'kokkosp-perfetto.pftrace'... Done

dalg24 reviewed (commented):
Looks good other than the couple comments

dalg24 commented:
Well looks like we still have the two branches so you need to target `develop`

jrmadsen committed:
endl instead of flush when finalizing

Co-authored-by: Damien L-G <dalg24+github@gmail.com>

jrmadsen committed:
Update profiling/perfetto-connector/Makefile

Co-authored-by: Damien L-G <dalg24+github@gmail.com>

jrmadsen committed:
Fix strip for macOS

- also fix unused const variable warning from perfetto

dalg24 reviewed (approved):
`strip` rule is being re-run every time but I can live with it.

I just need to check with @crtrott regarding the copyright for the Perfetto files before we merge.

crtrott commented:
Hey this looks good, trying to figure out the license situation. We may wanna add something like this statement from LLVM:

```
==============================================================================
Software from third parties included in the LLVM Project:
==============================================================================
The LLVM Project contains third party software which is under different license
terms. All such code will be identified clearly using at least one of two
mechanisms:
1) It will be in a separate directory tree with its own `LICENSE.txt` or
   `LICENSE` file at the top containing the specific license and restrictions
   which apply to that software, or
2) It will contain specific license and restriction terms at the top of every
   file.
```

But we will deal with it.

jrmadsen commented:
> `strip` rule is being re-run every time but I can live with it.

Not sure that can be fixed. Every time it runs, it touches libperfetto-connector.so so it thinks it's out-of-date on the next run.

dalg24 commented:
> > `strip` rule is being re-run every time but I can live with it.
> 
> Not sure that can be fixed. Every time it runs, it touches libperfetto-connector.so so it thinks it's out-of-date on the next run.

Could be part of the same rule that produce the library I suppose.  But no big deal.

jrmadsen committed:
Moved perfetto files into subdir and added LICENSE

jrmadsen commented:
@dalg24 @crtrott I moved perfetto.h and perfetto.cc into a perfetto subdirectory, added the LICENSE from their repo, and also copied it into the top of the header.

jrmadsen commented:
> Could be part of the same rule that produce the library I suppose. But no big deal.

Yeah I actually made it separate just in case, for some reason, someone wanted to build and not strip it. Although I guess they could just remove the line if the wanted

jrmadsen commented:
Ping

jrmadsen commented:
@crtrott @dalg24 is the license stuff resolved?

crtrott commented:
We have the license stuff resolved !! 

I.e. we finally got license change to Apache2 with LLVM exceptions approved, including the wording about containing code with other licenses in subdirectory.

crtrott commented:
We are currently doing the license update right now in core and kernels.

crtrott committed:
Apply clang-format

jrmadsen commented:
BTW, it should probably be noted in the chrome-tracing that this is effectively the newer version of that tool.

Do we want to apply formatting to the external perfetto files or exclude it from formatting?

crtrott committed:
Make perfetto compatible with new library build system

crtrott committed:
Add missing exposure of dual_view sync and modify

crtrott committed:
Apply clang-format

crtrott committed:
Disable clang-format inside of perfetto itself

crtrott committed:
Fix format
